x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java

```diff
@@ -265,6 +475,74 @@ public static Hasher resolve(String name, Hasher defaultHasher) {
         }
     }
 
+    public static Hasher resolveFromHash(char[] hash) {
+        String hashString = new String(hash);
+        if (hashString.startsWith(BCRYPT_PREFIX)) {
+            return Hasher.BCRYPT;
+        } else if (hashString.startsWith(PBKDF2_PREFIX)) {
+            return Hasher.PBKDF2;
+        } else if (hashString.startsWith(SHA1_PREFIX)) {
+            return Hasher.SHA1;
+        } else if (hashString.startsWith(MD5_PREFIX)) {
+            return Hasher.MD5;
+        } else if (hashString.startsWith(SSHA256_PREFIX)) {
+            return Hasher.SSHA256;
+        } else {
+            throw new IllegalArgumentException("unknown hash format for hash [" + hashString + "]");
+        }
+    }
+
+    public static boolean verifyHash(SecureString data, char[] hash) {
+        try {
+            final Hasher hasher = resolveFromHash(hash);
+            return hasher.verify(data, hash);
+        } catch (IllegalArgumentException e) {
+            // The password hash format is invalid, we're unable to verify password
+            return false;
+        }
+    }
+
+    private static char[] getPbkdf2Hash(SecureString data, int cost) {
+        try {
+            // Base64 string length : ((4*n/3)+3) rounded up to the next multiple of 4 because of padding = 44
+            CharBuffer result = CharBuffer.allocate(PBKDF2_PREFIX.length() + String.valueOf(cost).length() + 2 + 32 + 44);
+            result.put(PBKDF2_PREFIX);
+            result.put(String.valueOf(cost));
+            result.put("$");
+            char[] salt = SaltProvider.salt(32);  // 32 characters for 64 bytes
+            result.put(salt);
+            result.put("$");
+            SecretKeyFactory secretKeyFactory = SecretKeyFactory.getInstance("PBKDF2withHMACSHA512");
+            PBEKeySpec keySpec = new PBEKeySpec(data.getChars(), CharArrays.toUtf8Bytes(salt), cost, 256);
+            result.put(Base64.getEncoder().encodeToString(secretKeyFactory.generateSecret(keySpec).getEncoded()));
+            return result.array();
+        } catch (InvalidKeySpecException | NoSuchAlgorithmException e) {
+            throw new IllegalStateException("Can't use PKDF2 for password hashing", e);
```

Comment:
nit: PBKDF2
I think you can throw `ElasticSearchException`. I think this is the goto unchecked exception unless smth else fits better.